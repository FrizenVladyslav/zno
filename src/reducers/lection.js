import clone from 'lodash/clone'

const prefix = 'LECTION_'

export const actionTypes = {
  ADD: `${prefix}ADD`,
  EDIT: `${prefix}EDIT`,
  GET: `${prefix}GET`,
  SET: `${prefix}SET`,
  UNSET: `${prefix}UNSET`,
}

const initState = {
  lections: null,
  lection: null,
}

export default function lectionReducer(state = initState, action) {
  switch (action.type) {
    case actionTypes.ADD:
      return {
        ...state,
        lections: [...state.lections, action.payload],
      }
    case actionTypes.EDIT:
      const { id } = action.payload
      const lections = clone(state.lections)
      const index = lections ? lections.findIndex(({ _id }) => _id === id) : -1
      if (index !== -1) {
        const updatedLection = Object.assign(lections[index], action.payload)
        lections[index] = updatedLection
        return {
          ...state,
          lections,
        }
      } else return state
    case actionTypes.GET:
      return {
        ...state,
        lection: action.payload,
      }
    case actionTypes.SET:
      return {
        ...state,
        lections: action.payload,
      }
    case actionTypes.UNSET:
      return initState
    default:
      return state
  }
}
