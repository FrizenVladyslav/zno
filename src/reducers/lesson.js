import clone from 'lodash/clone'

const prefix = 'LESSON_'

export const actionTypes = {
  ADD: `${prefix}ADD`,
  EDIT: `${prefix}EDIT`,
  SET: `${prefix}SET`,
}

const initState = {
  lessons: [],
}

export default function lessonReducer(state = initState, action) {
  switch (action.type) {
    case actionTypes.ADD:
      return {
        ...state,
        lessons: [...state.lessons, action.payload],
      }
    case actionTypes.EDIT:
      const { id, title } = action.payload
      const lessons = clone(state.lessons)
      lessons[lessons.findIndex(({ _id }) => _id === id)].title = title

      return {
        ...state,
        lessons,
      }
    case actionTypes.SET:
      return {
        ...state,
        lessons: action.payload,
      }
    default:
      return state
  }
}
