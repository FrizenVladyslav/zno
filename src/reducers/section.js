import clone from 'lodash/clone'

const prefix = 'SECTION_'

export const actionTypes = {
  ADD: `${prefix}ADD`,
  EDIT: `${prefix}EDIT`,
  SET: `${prefix}SET`,
  UNSET: `${prefix}_UNSET`,
}

const initState = {
  sections: null,
}

export default function sectionReducer(state = initState, action) {
  switch (action.type) {
    case actionTypes.ADD:
      return {
        ...state,
        sections: [...state.sections, action.payload],
      }
    case actionTypes.EDIT:
      const { id, title } = action.payload
      const sections = clone(state.sections)
      sections[sections.findIndex(({ _id }) => _id === id)].title = title

      return {
        ...state,
        sections,
      }
    case actionTypes.SET:
      return {
        ...state,
        sections: action.payload,
      }
    case actionTypes.UNSET:
      return { ...initState }
    default:
      return state
  }
}
