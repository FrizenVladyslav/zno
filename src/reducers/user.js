const prefix = 'USER_'

export const actionTypes = {
  SET: `${prefix}SET`,
  UNSET: `${prefix}UNSET`,
}

const initState = {
  role: 'Anon',
}

export default function userReducer(state = initState, action) {
  switch (action.type) {
    case actionTypes.SET:
      return action.payload
    case actionTypes.UNSET:
      return initState
    default:
      return state
  }
}
