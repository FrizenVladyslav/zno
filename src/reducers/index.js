import { combineReducers } from 'redux'

import lection from './lection'
import lesson from './lesson'
import section from './section'
import user from './user'

export default combineReducers({
  lection,
  lesson,
  section,
  user,
})
