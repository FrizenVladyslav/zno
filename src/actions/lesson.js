import Api from './api'
import store from 'utils/store'
import { actionTypes } from 'reducers/lesson'

const endpoint = 'lesson'

export async function add(title) {
  let res = await Api.post(endpoint, { title })
  if (res.status !== 201) throw new Error('Lesson not created')

  store.dispatch({ type: actionTypes.ADD, payload: res.data })
}

export async function get() {
  let res = await Api.get(endpoint)
  if (res.status !== 200) throw new Error('Lessons not geted')

  store.dispatch({ type: actionTypes.SET, payload: res.data })
}

export async function edit(id, title) {
  let res = await Api.put(`${endpoint}/${id}`, { title })
  if (res.status !== 200) throw new Error('Lesson not edited')

  store.dispatch({ type: actionTypes.EDIT, payload: { id, title } })
}
