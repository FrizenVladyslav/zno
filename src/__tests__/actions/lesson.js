// @ts-nocheck
import configureStore from 'redux-mock-store'
import { actionTypes } from 'reducers/lesson'

const storeMock = configureStore()
const store = storeMock()

describe('LessonActions', () => {
  afterEach(() => {
    store.clearActions()
  })

  it('add lesson action', () => {
    store.dispatch({ type: actionTypes.ADD, payload: { title: 1 } })
    expect(store.getActions()).toMatchSnapshot()
  })

  it('get lessons action', () => {
    store.dispatch({ type: actionTypes.SET, payload: [1, 2, 3] })
    expect(store.getActions()).toMatchSnapshot()
  })
})
