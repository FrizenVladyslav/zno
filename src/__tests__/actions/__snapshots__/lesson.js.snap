// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LessonActions add lesson action 1`] = `
Array [
  Object {
    "payload": Object {
      "title": 1,
    },
    "type": "LESSON_ADD",
  },
]
`;

exports[`LessonActions add lessons action 1`] = `
Array [
  Object {
    "payload": Array [
      1,
      2,
      3,
    ],
    "type": "LESSON_ADD",
  },
]
`;

exports[`LessonActions get lessons action 1`] = `
Array [
  Object {
    "payload": Array [
      1,
      2,
      3,
    ],
    "type": "LESSON_SET",
  },
]
`;
